check if allocated dedicated host is available (with timeout) - fixes #216

hidalgopl opened the pull request:
There are two issues that this should fix:
1. If there's no dedicated host available, kip tried to allocate a new one and run an instance on it without check if it's available.
2. If there was an available host but fully occupied, kip tried to run instance on it, as it was checking only `state`. Should be fixed now.

I'm still getting
```
kip-provider-0 kip I0126 14:51:01.640099       1 node_controller.go:327] Unhealthy wait for running: waiting for instance to start: ResourceNotReady: failed waiting for successful resource state, terminating node: 31ecc3c4-b4b8-47b2-9fd0-52d94514ebd8
```
from 
  `e.client.WaitUntilInstanceRunning(dii)`. I tried to increase retry timeout, but it didn't help.

jrroman commented:
> There are two issues that this should fix:
> 
>     1. If there's no dedicated host available, kip tried to allocate a new one and run an instance on it without check if it's available.
> 
>     2. If there was an available host but fully occupied, kip tried to run instance on it, as it was checking only `state`. Should be fixed now.

I don't know that 2 is actually an issue within the implementation of requesting and allocating hosts we are only returning hosts that are not currently occupied. Im not sure as to where you are seeing that we are only checking for 'state'?

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -17,7 +17,9 @@ limitations under the License.
 import (
+	"context"
 	"fmt"
+	"github.com/elotl/kip/pkg/server/nodemanager"
```
This is a no-no: even though it's not an explicit circular dependency, the node controller uses the cloud client which is implemented by this package for AWS, so it's pretty close. It's 100x better to hardcode a timeout value than to introduce a circular dependency.

hidalgopl replied:
good point :+1: I forgot to copy this var to the aws package.

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -455,6 +496,12 @@ func (e *AwsEC2) StartDedicatedNode(node *api.Node, image cloud.Image, metadata,
 	volSizeGiB := cloud.ToSaneVolumeSize(node.Spec.Resources.VolumeSize, image)
 	devices := e.getBlockDeviceMapping(image, volSizeGiB)
 	networkSpec := e.getInstanceNetworkSpec(node.Spec.Resources.PrivateIPOnly)
+	ctx, _ := context.WithTimeout(context.Background(), 30 * time.Second)
```
The second return value should be called before returning this function to prevent a resource leak.

hidalgopl replied:
:+1: added explicit `cancel()` before sending `RunInstance` request

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -441,6 +450,38 @@ func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
+				return nil
+			}
+		default:
+			hostAvailable = e.isHostAvailable(hostId)
```
Should not this return once the host is available, to prevent unnecessary waiting?

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -441,6 +450,38 @@ func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
+		default:
+			hostAvailable = e.isHostAvailable(hostId)
+		}
+		time.Sleep(500 * time.Millisecond)
```
Instead of a direct sleep here, you could use a Ticker in the `select` statement above instead of `default`.

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -441,6 +450,38 @@ func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
+		return false
+	}
+	host := describeOutput.Hosts[0]
+	return *host.State == "available"
```
Please use `aws.StringValue()` to check the value safely.

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -441,6 +450,38 @@ func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
 
+func (e *AwsEC2) isHostAvailable(hostId *string) bool {
+	describeOutput, err := e.client.DescribeHosts(&ec2.DescribeHostsInput{
+		HostIds:    []*string{hostId},
```
The `aws` package has a function for create a slice of string pointers, IIRC `aws.StringSlice()`.

ldx reviewed (commented):
Thanks for fixing this! I've dropped a few comments with suggestions.

hidalgopl commented:
FYI
I think that the issue we're hitting is related with volume size of the AMI. We started seeing it after we resized attached volume to 500Gb. After my changes from this PR, kip gets to the point when it sends `RunInstances` request, but instance creation fails and dedicated host almost immediately turns its state to Pending.
![Screenshot from 2021-01-27 19-13-32](https://user-images.githubusercontent.com/18266530/106035153-0ea59d00-60d4-11eb-9d03-67d48cc37bd6.png)
As you see, there's `Client.InvalidParameterCombination: Could not create volume with size 500GiB from snapshot 'snap-064acde5cd376cb6b'`. I've added logging block devices volume size to debug if correct numbers are passed to `RunInstances` function call.
```
kip-provider-0 kip I0127 17:45:33.909398       1 instances.go:493] Starting instance for node: &{{Node v1} {26b5749b-a378-4a77-9ece-05314c889606 map[] 2021-01-27 17:45:33.90770785 +0000 UTC <nil> map[] 08358c1e-5551-4f58-be7a-c4119665a587 default} {mac1.metal ami-0929815870cdeaa46 false false true {   20G false <nil> false <nil>}} {Creating  [] default_buildkite-agent-mac1-metal}}
kip-provider-0 kip I0127 17:45:34.209321       1 instances.go:504] calculated volume size for node: 500
kip-provider-0 kip I0127 17:45:35.209502       1 instances.go:480] checking host h-020fba8d6384e6cc9 availability...
kip-provider-0 kip I0127 17:45:35.281221       1 instances.go:514] Starting node with security groups: [sg-0ee086454488a7451] subnet: 'subnet-2769f140'
kip-provider-0 kip I0127 17:45:35.281250       1 instances.go:516] Block devices for a node
kip-provider-0 kip I0127 17:45:35.281257       1 instances.go:518] Device: /dev/sda1 volume size: 824658735208
kip-provider-0 kip I0127 17:45:36.707965       1 instances.go:552] Started instance: i-01b2507ae043e4275
kip-provider-0 kip I0127 17:45:51.843890       1 instances.go:642] retrying err: ResourceNotReady: failed waiting for successful resource state
``` 
there's probably an issue with the volume size, as you may see from logs

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -462,23 +466,25 @@ func (e *AwsEC2) isHostAvailable(hostId *string) bool {
+				quit <- true
+			}
+		case <-quit:
+			return hostAvailable
```
This could be simplified further if you removed `quit`, and return `hostAvailable` directly when it's true after the `isHostAvailable()` check.

hidalgopl commented on pkg/server/nodemanager/node_scaler.go:
```diff
@@ -102,8 +102,9 @@ func (s *BindingNodeScaler) createNodeForPod(pod *api.Pod) *api.Node {
+
 	node.Spec.Resources.VolumeSize = s.defaultVolumeSize
-	if s.fixedSizeVolume && pod.Spec.Resources.VolumeSize != "" {
+	if pod.Spec.Resources.VolumeSize != "" { // TODO: consider if && s.fixedSizeVolume is needed here
```
@ldx @henryprecheur for AWS `s.fixedSizeVolume` is always set to false. Which basically means that pod annotation `pod.elotl.co/volume-size: "X Gi"` has no effect, which may be confusing for user (it isn't also specified nowhere in docs for which cloud providers this annotation works). I needed to manually set volume size to test if the size is an issue and noticed that. Should we address it in a separate issue or is this desired behavior?

ldx replied:
This annotation does work on AWS. However, resizing happens in the pod controller, not in the node controller when `fixedSizeVolume` is false (i.e. on AWS).

hidalgopl commented:
Ok, so there were few issues here:
1. Retrying timeout for checking if instance is available was too small - I increased it 
2. We had hardcoded volume type for all EBSes - `gp2`. I've added getting volume type from AMI's volumes and using it.
3. We were checking only for host state (if it's available) without checking if it has free compute (which result in attempts of trying to run an instance on the occupied host) - I've added additional check if it has available vCPUs.
4. We haven't been checking newly allocated host state - added additional check.

ldx commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -441,6 +466,41 @@ func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
+				return hostAvailable
+			}
+			klog.V(2).Infof("host %s not available yet", *hostId)
+		default:
```
Don't leave in the `default` case, otherwise it will always be taken while the other two cases are not true, effectively busy looping here.

hidalgopl replied:
:+1: removed

ldx commented on pkg/server/nodemanager/node_controller.go:
```diff
@@ -324,7 +324,6 @@ func (c *NodeController) finishNodeStart(node *api.Node) error {
 	// a describe instance here...
 	addresses, err := c.CloudClient.WaitForRunning(node)
 	if err != nil {
-		klog.V(2).Infof("Unhealthy wait for running, terminating node: %s", node.Name)
```
Just double-checking that you wanted to remove this log line and we don't need it for diagnosing node start problems?

hidalgopl replied:
Good catch! I went too far in removing debug lines added by me :+1: reverted

ldx reviewed (approved):
:+1:

henryprecheur commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -441,6 +466,41 @@ func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
+	return aws.StringValue(host.State) == AwsInstanceAvailableState
+}
+
+func (e *AwsEC2) waitForHostAvailable(ctx context.Context, hostId *string) bool {
```
This function needs a timeout. Otherwise it may wait for the host to come up forever.

ldx replied:
> This function needs a timeout. Otherwise it may wait for the host to come up forever.

The caller specifies a timeout via `ctx`.

hidalgopl commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -40,6 +41,11 @@ const (
 	elotlImageNameFilter  = "elotl-kip-*"
 )
 
+var (
```
moved to `consts` block :+1:

hidalgopl commented on pkg/server/cloud/aws/instances.go:
```diff
@@ -274,6 +280,16 @@ func getRootDeviceVolumeSize(blockDevices []*ec2.BlockDeviceMapping, rootDeviceN
+		}
+	}
+	return volumeType
+}
```
I like the idea, merged those two :+1:
